mage/README.md

```diff
@@ -0,0 +1,53 @@
+# Mage
+This is a highly experimental build framework that writes the complicated build logic in golang. This allows to write
+the build system in the same language as the project itself, take advantage of any libraries and share code between
+projects, and more aggresively take advantage of caching and parralelization to get out of devs way.
+
+Due to the complexity of the current Makefile I'm not confident I've fully replicated them in go, however, the outcome
+of the build can be installed.
+
+## Getting Started
+
+### OSS
+
+`go run mage.go cluster:gravity`
+
+This is roughly equivelant to `make production telekube` within the OSS source directory. 
+
+### Enterprise
+
+`ENTERPRISE=true go run mage.go cluster:gravity`
+
+This is roughly equivelant to `make production telekube` within the enterprise source directory. 
+
+## Configuration
+Mage doesn't support build arguments, so all customization of the build process is done via environment variables. For
+a complete listing of environment variables, run the help:envs target. This will give a listing of all env variables
+and any current settings.
+
+`go run mage.go help:envs`
+
+## Notes and Tips
+
+1. There isn't really a good way to rebuild specific packages bypassing caching, besides changing the version that
+would be assigned. So for example to force rebuild RBAC would require changing the version by env variable.
```

Comment:
This is in line with caching behaviour today, but a little more exposed. Even today we try and be clever with some caching, which is effectively controlled by the version number. So if rbac-app:1.0.0 has been built before and uploaded to AWS, if we're able to download rbac-app:1.0.0 it will use that instead of rebuilding it, even if the contents have changed. 

The difference here is I don't try and download from AWS at all, but still cache locally by the version number. So if you build rbac-app:1.0.0-wadells.test.1, make some changes, and build again, the resulting installer may have the cached/old version. 

The way I'd imagine this working in the short term would by to do something like:
`FORCE_APP_REBUILD=true go run mage.go package:rbac && go run mage.go cluster:gravity` 

This would force rebuild some app being worked on, allow all other apps to be served from cache, but haven't gotten around to implementing it. For now it means changing the tag every build to avoid caches, so rbac-app:1.0.0-wadells.test.2, test.3, etc. 

Ideally we'd be able to use BuildKit as an Inputs cache, but as I explained on the last team call one of the problems we have is that the gravity binary is an input to the content cache, meaning each change to the gravity binary invalidates the caches. We don't change the format very frequently, so we might be able to get to a point where we can do something here, or maybe it would be better to just add content caching directly to our tools. 

Changing the tag to build though on the other hand is a good practice. Gravity internally uses the version number to decide to upgrade, so it has happened more than once where contents have been changes, but the version number has been missed.